Add Hardware CRC HAL API specification

scartmell-arm opened the pull request:
### Description

Define the HAL API header for the Hardware CRC module. This set of functions
allows hardware acceleration of a subset of CRC algorithms for supported
platforms by providing access to the hardware CRC module of certain platforms.

### Status

NOT READY

### Tasks

[X] Integrate the HAL API into the Platform API
[ ] Add Reference Implementation for STM32X
[X] Add Reference Implementation for K64F
[ ] Add Reference Implementation for Maxim 32600

### Pull request type

[ ] Fix  
[ ] Refactor  
[ ] New target  
[X] Feature  
[ ] Breaking change


@bulislaw  @deepikabhavnani

Commit:
Add Hardware CRC HAL API specification headers

Define the HAL API header for the Hardware CRC module. This set of functions
allows hardware acceleration of a subset of CRC algorithms for supported
platforms by providing access to the hardware CRC module of certain platforms.

The API is defined as four separate functions:

  - hal_crc_is_supported(polynomial)

    Indicates to the caller if the specific CRC polynomial is supported.

  - hal_crc_compute_partial_start(const uint32_t polynomial)

    Initializes the hardware CRC module with the given polynomial.

  - hal_crc_compute_partial(*data, size)

    Writes an array of bytes to the CRC module to be appended to the calculation

  - hal_crc_get_result()

    Applies the final transformations to the data and returns the result to the
    caller.

Commit:
Add K64F Hardware CRC reference implementation


- Add support for all currently defined polynomials
- Add 'CRC' flag to targets.json for K64F profile

Commit:
Adapt MbedCRC class to support Hardware CRC 

- Move CRC polynomial enum into HAL layer, so it's accessible from platform
  implementations

- Add enum to CRC class to indicate which mode the CRC class should use:
  HARDWARE, TABLE, or BITWISE

- Add calls to HAL Hardware CRC API to each of the compute functions when the
  class is in HARDWARE mode.

- Add missing constructor call to template constructor, and remove const from
  delegating constructor.

bulislaw commented on drivers/MbedCRC.h:
```diff
@@ -247,6 +262,7 @@ class MbedCRC
     bool _reflect_data;
     bool _reflect_remainder;
     uint32_t *_crc_table;
+    crc_mode mode_;
```
`mode_` -> `_mode`

bulislaw commented on hal/crc_api.h:
```diff
@@ -0,0 +1,130 @@
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
```
As with other HAL APIs we need some intro (have a look at ticker/rtc/sleep)

bulislaw commented on hal/crc_api.h:
```diff
@@ -0,0 +1,130 @@
+ * from two different contexts may lead to configurations being overwrite and
+ * results being lost.
+ *
+ * \param polynomial CRC Polynomial. Example polynomial: 0x1021 = x^12+x^5+1
```
Link the enum.

mprse commented on hal/crc_api.h:
```diff
@@ -0,0 +1,117 @@
+#ifndef MBED_CRC_HAL_API_H
```
I think that the header file needs to be updated to be consistent with other HAL headers.
Doxygen groups, requirements, undefined behavior, potential bugs sections should be added. Example can be found here:
https://github.com/ARMmbed/mbed-os/blob/feature-hal-spec-ticker/hal/us_ticker_api.h

I have also created some example update with the requirements for testing, which can be found here:
https://github.com/mprse/mbed-os/blob/scartmell-arm_crc_tests/hal/crc_api.h

The requirements were created based on provided function descriptions, so probably they need to be extended. 
Please fill free to use this example in order to update the header file.

mprse commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/mbed_crc_api.c:
```diff
@@ -0,0 +1,125 @@
+    }
+    case POLY_8BIT_CCITT:
+    {
+      width = kCrcBits16;
```
Why `width` is set to 16-bits if 8-bit CRC is requested?

mprse replied:
If hardware can handle only 16 or 32 bits CRC, then `POLY_8BIT_CCITT`, `POLY_7BIT_SD` polynomials should be not supported by this platform.

mprse replied:
Maybe I'm missing something, but according to K64F reference manual:

> The cyclic redundancy check (CRC) module generates 16/32-bit CRC code for error
> detection.

It looks like currently `POLY_8BIT_CCITT` and `POLY_7BIT_SD` generates 16 bit CRC. We expect to have 8/7 bit CRC. `config.crcBits` is set in both cases to `kCrcBits16` (and on K64F platform we don't have definitions for 7/8 bits CRC). Only different polynomials are used in each case.

Please look at the following test:
```
    uint32_t crc = 0;
    const uint8_t test_data[] = "123456789";
    hal_crc_compute_partial_start(POLY_8BIT_CCITT);
    hal_crc_compute_partial((uint8_t*) test_data, strlen((const char*)test_data));
    crc = hal_crc_get_result();

    printf("Result: 0x%X \n", crc);
    printf("Expected: 0x%X \n", 0xF4);

output:
Result: 0xEF6F  <--- 16 bit result
Expected: 0xF4

I took expected result (0xF4) from `tests-mbed_drivers-crc`.
```

Commit:
Fix CRC class coding conventions

deepikabhavnani commented on hal/crc_api.h:
```diff
@@ -0,0 +1,160 @@
+ * \param data Input data stream to be written into the CRC calculation
+ * \param size Size of the data stream in bytes
+ */
+void hal_crc_compute_partial(const uint8_t *data, const size_t size);
```
We do not get any failures from hardware? I believe return of computation should be non-zero and error codes should be propagated

deepikabhavnani commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/mbed_crc_api.c:
```diff
@@ -0,0 +1,101 @@
+    case POLY_16BIT_CCITT:
+    case POLY_16BIT_IBM:
+    case POLY_32BIT_ANSI:
+      return true;
```
K64F supports only specified polynomials or any polynomial with width 16/32?
If we have  hal API to set configs, user can select any polynomial and not just above three.

deepikabhavnani commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/mbed_crc_api.c:
```diff
@@ -0,0 +1,101 @@
+      break;
+    }
+    default:
+      MBED_ASSERT("Configuring Mbed CRC with unsupported polynomial");
```
Provide an option to set these configs and not ASSERT, it should be user configurable to have 16BIT_CCIIT but seed as Zero. I am sure it will work perfectly fine with HW.

deepikabhavnani replied:
You can overload functions to set those parameters. I don;t think sticking to only specified polynomials make sense for HW. In case of software we need to add multiple tables for each polynomial which is not the case in HW.
Also with the same polynomial even SW CRC gives option to have different seed / reflection settings. How are you planning to support this constructor https://github.com/ARMmbed/mbed-os/blob/master/drivers/MbedCRC.h#L127

Also I have strong use case for this requirement, SD driver uses standard CCITT 16-bit but with seed value of 0. https://github.com/ARMmbed/sd-driver/blob/master/SDBlockDevice.cpp#L251

deepikabhavnani replied:
I don;t think sticking to only specified polynomials make sense for HW.  - Unless HW supports only those few

bulislaw replied:
The CRC support in HW is rather modest. I don't really like the idea of complicating the API for K64F only for CRC polynomials that are not a very standard ones. Where the HW support actually matters is when we need to calculate the checksum repeatedly and fast eg network packets, bootloader maybe, but that should be using standard polynomials.

deepikabhavnani replied:
@bulislaw - Polynomials supported in software CRC are not the only standard ones. Please see the list on Wiki "https://en.wikipedia.org/wiki/Cyclic_redundancy_check" for list.

Commit:
Remove support for 7/8 bit CRC polynomials for K64F

mprse commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/mbed_crc_api.c:
```diff
@@ -0,0 +1,125 @@
+
+void hal_crc_compute_partial(const uint8_t *data, const size_t size)
+{
+  CRC_WriteData(CRC0, data, size);
```
I suggest to add robust checks against null pointer and 0 size and add the following information in the requirements section:

`Function hal_crc_compute_partial() does nothing if pointer to buffer is undefined (NULL) or data length is equal to 0`.

On different platforms data size equal to 0 might have different consequences, so we can handle such case and simply return without any action.

Commit:
HAL CRC: Return early when calling compute with null or 0 size buffer

theotherjimmy reviewed (changes requested):
You defined a subset of what the standard supports: you don't include seed as a parameter, yet you also do not even use the same seed as managed boot loader mode for CRCITT32: 0.

Support the seed parameter.

Commit:
Add CRC configuration options to HAL API

mprse commented on hal/crc_api.h:
```diff
@@ -35,6 +35,19 @@ typedef enum crc_polynomial {
     POLY_32BIT_ANSI  = 0x04C11DB7, // x32+x26+x23+x22+x16+x12+x11+x10+x8+x7+x5+x4+x2+x+1
 } crc_polynomial_t;
 
+typedef struct crc_mbed_config {
```
I see that conception has changed. This change has massive impact on my test, so please provide requirements for testing.
I have also few questions, so I can start re-factoring my test.
Do we assume now that CRC platforms must support all of the listed options?
Why CRC width is not specified in `crc_mbed_config` (we can have different CRC width for the same polynomial)?
Does conception with predefined typical polynomials is eventually abandoned?

mprse replied:
Thanks for the answers. I have two more questions:
1. Do we still limit the number of supported polynomials to ones provided in `crc_polynomial_t` enum with additional config options (initial_xor, final_xor, reflect_in, reflect_out).
2. Why initial value config option is not available?

mprse replied:
> This is the same as an initial_xor?

Looks like it is :-) Thanks.

mprse commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/mbed_crc_api.c:
```diff
@@ -0,0 +1,65 @@
+  if (config == NULL)
+    return false;
+
+  if ((config->polynomial & 0x80008000) == 0)
```
It look like this check is to return false in case when CRC size is different than CRC16 or CRC32 - only these sizes can be handled by K64F hardware. 
Please check the predefined polynomial values:
```
POLY_16BIT_CCITT = 0x1021
POLY_32BIT_ANSI  = 0x04C11DB7
````

In both cases `hal_crc_is_supported` function will return `false`. I still suggest to add CRC size field into `crc_mbed_config` structure.

mprse commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/mbed_crc_api.c:
```diff
@@ -0,0 +1,65 @@
+  platform_config.seed               = config->initial_xor;
+  platform_config.reflectIn          = config->reflect_in;
+  platform_config.reflectOut         = config->reflect_out;
+  platform_config.complementChecksum = true;
```
If given `final_xor` config value is equal to `all ones` (e.q. 0xFFFF_FFFF for 32-bit CRC), then `platform_config.complementChecksum` should be set to `true`. 
If given `final_xor` config value is 0, then `platform_config.complementChecksum` should be set to `false`.
If given `final_xor` config value is different than both mentioned values, then `platform_config.complementChecksum` should be set to `false`, software `xor` operation should be performed on the hardware result and the result of the operation should be returned.

deepikabhavnani replied:
Complement checksum is different name for final XOR. Its same parameter but different hardware modules accept either true or full value. It should not be separate argument from HAL.

mprse commented on hal/crc_api.h:
```diff
@@ -0,0 +1,176 @@
+ */
+typedef enum crc_polynomial {
+    POLY_OTHER       = 0,
+    POLY_8BIT_CCITT  = 0x07,       // x8+x2+x+1
```
Why `x8` has been added in the comment?
`0x07 == 0000 0111b => x^2 + x^1 + x^0 = x^2 + x + 1`

Next definitions also have this issue.

Commit:
HAL CRC: Add bit width parameter to crc_config_t

0xc0170 commented on hal/crc_api.h:
```diff
@@ -0,0 +1,178 @@
+} crc_polynomial_t;
+
+typedef struct crc_mbed_config {
+    /// CRC Polynomial. Example polynomial: 0x21 = 0010_0011 = x^5+x+1
```
Good to have docs, but these should be done as `/**< text > */` as some others in hal folder

0xc0170 commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/mbed_crc_api.c:
```diff
@@ -0,0 +1,65 @@
+#include "crc_api.h"
```
missing license header on top

0xc0170 commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/mbed_crc_api.c:
```diff
@@ -0,0 +1,65 @@
+bool hal_crc_is_supported(const crc_mbed_config_t* config)
+{
+  if (config == NULL)
+    return false;
```
Please review coding style (hoewever targets/ are not checked), there are others that shall be fixed

Commit:
Add support for final_xor for K64F Hardware CRC

deepikabhavnani commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/mbed_crc_api.c:
```diff
@@ -0,0 +1,72 @@
+  if (config == NULL)
+    return;
+
+  width = ((config->polynomial & 0xFFFF0000U) != 0) ? kCrcBits32 : kCrcBits16;
```
Why do we need this extra intelligence here? User passes the width, and it is available as part of config->width. We can simply use that.

Commit:
HAL CRC: Fix code to match coding conventions

0xc0170 commented on hal/crc_api.h:
```diff
@@ -0,0 +1,178 @@
+ *
+ * \param polynomial CRC Polynomial. Example polynomial: 0x1021 = x^12+x^5+1
+ */
+void hal_crc_compute_partial_start(const crc_mbed_config_t* polynomial);
```
HAL API has first parameter this pointer (*obj), not needed for this API? This is hardware related functionality, thus will be needed to have access to their SDK or internal registers (K64F below has CRC0).

0xc0170 replied:
The point was that all HAL API have defined as first parameter "this" pointer. My question is - why these new functions do not have it

bulislaw replied:
We don't need it, there is no platforms with multiple CRC engines (no real need for that) and there's no state/context that we need to be maintained.

deepikabhavnani replied:
Some HW CRC modules do store intermediate CRC state, which might be an issue when using CRC HW from different threads.

bulislaw replied:
Driver is providing the locking, HAL specification explicitly mentions that it's up to the user to ensure access safety. HAL is designed for simplicity, the drivers add the level of complexity necessary for safe usage for the user.

Commit:
HAL CRC: K64F: Set output width based on input width, not the polynomial

Commit:
HAL CRC: Add DEVICE_CRC macro to doxygen files

Commit:
HAL CRC: Update function documentation with new config parameter

bulislaw commented:
@scartmell-arm let's try to merge it ASAP and add extra platforms after that.

@theotherjimmy @0xc0170 @deepikabhavnani we are getting close to code freeze can you guys rereview the code and accept it if your comments were resolved.

theotherjimmy commented:
@bulislaw I would like to approve this, but I don't see any indication that it's thread safe, nor to I see documentation that It's not thread safe.

deepikabhavnani commented on drivers/MbedCRC.h:
```diff
@@ -408,9 +432,25 @@ class MbedCRC
     {
         MBED_STATIC_ASSERT(width <= 32, "Max 32-bit CRC supported");
+
+        _mode = (_crc_table == NULL) ? TABLE : BITWISE;
```
Logic here seems reverted, If _crc_table is NULL then BITWISE computation should be done. 
https://github.com/ARMmbed/mbed-os/pull/6708/files#diff-a353a07f9612be80e818a21f61ec8409L181****

deepikabhavnani commented on hal/crc_api.h:
```diff
@@ -0,0 +1,182 @@
+ * \return The final CRC checksum after the reflections and final calculations
+ *         have been applied.
+ */
+uint32_t hal_crc_get_result(void);
```
None of the CRC functions are thread safe, please indicate in this API as well

deepikabhavnani commented on hal/crc_api.h:
```diff
@@ -0,0 +1,182 @@
+ *
+ * \return  True if running if the polynomial is supported, false if not.
+ */
+bool hal_crc_is_supported(const crc_mbed_config_t* config);
```
Good to inform here as well, that it is not thread safe.

deepikabhavnani commented:
Do we have test PR in progress for this, Can we link here

Commit:
HAL CRC: Fix inverted CRC mode selection

Commit:
HAL CRC: Add additional thread safety notes to the API

deepikabhavnani reviewed (dismissed):
👍 Looks good

ghost commented:
> Do we have test PR in progress for this, Can we link here

https://github.com/ARMmbed/mbed-os/pull/6831

Commit:
HAL CRC: Basic implementation for STM32F0

ghost commented:
/morph build

mprse commented on targets/TARGET_STM/TARGET_STM32F0/mbed_crc_api.c:
```diff
@@ -0,0 +1,69 @@
+    return false;
+  }
+
+  if (config->polynomial != DEFAULT_CRC32_POLY) {
```
Shouldn't we use HAL definitions if possible? It looks like DEFAULT_CRC32_POLY corresponds POLY_32BIT_ANSI.
Please note that while testing the CRC HAL API I can not use target specific definitions. So I think that it is a good assumption that at least one of the predefined polynomials must be supported.

Commit:
HAL CRC: Use HAL polynomial enum instead of STM32 enum

Commit:
HAL CRC: Move STM32 implementation to common folder and add targets

ghost commented:
/morph build

0xc0170 commented:
@AnotherButler Can you review hal/crc_api.h docs ? 

@scartmell-arm Is there any docs update needed for this new HAL? If yes, please reference it here

0xc0170 commented:
I missed the reference above (505 PR in the handbook).

Final reviews , needs green checks !

Commit:
HAL CRC: Add additional STM32 targets

bulislaw reviewed (dismissed):
Looks good.

adbridge commented:
@deepikabhavnani @mprse Are you happy with the changes ?

cmonr commented:
@deepikabhavnani @mprse Fyi, this PR is holding up https://github.com/ARMmbed/mbed-os/pull/6831

0xc0170 commented:
/morph build

0xc0170 commented:
@scartmell-arm Failures look related, please review

ghost commented:
/morph build

bulislaw commented:
That's some crazy failures...

cmonr commented:
/morph build

0xc0170 commented:
> That's some crazy failures...

Correct, no space left. should have been resolved by now.

cmonr commented:
@bulislaw @theotherjimmy @0xc0170 @deepikabhavnani @mprse @AnotherButler Final re-review?

Commit:
Copy edit crc_api.h

Copy edit file, mostly for American spelling and consistent tense across files.

cmonr commented:
Making sure, are the STM32X and Maxim 32600 tasks not going to be completed for this PR?

cmonr commented:
/morph build

bulislaw commented:
STM is there, Maxim will be delivered later.

cmonr commented:
Performing a PR shuffle. Attempting to make sure https://github.com/ARMmbed/mbed-os/pull/6993 is ready by the AM, if needed.

/morph export-build

0xc0170 commented:
/morph uvisor-test

0xc0170 commented:
/morph uvisor-test

cmonr commented:
Well poop.
Gonna need a rebase to fix the target.json.

Will start CI after it's fixed.

ghost commented:
Fixed merge conflict.

cmonr commented:
/morph build

cmonr commented:
CI nodes want to leave for the weekend a bit too soon...
/morph export-build

0xc0170 commented:
This is now almost ready, but HAL #7009 should land first . Preceding PR label set

cmonr commented:
Swapping this PR's priority with #7009. 7009 will need updates before it becomes ready to merge, whereas this is already ready to go.
